### PR TITLE
OpenShift auth handler for v2 docker registry

### PR DESCRIPTION
**Pull request by pravisankar**

### CODE REVIEW

File: pkg/dockerregistry/auth/openshift.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+	}, nil
+}
+
+// Error returns the internal error string for this authChallenge.
+func (ac *authChallenge) Error() string {
+	return ac.err.Error()
+}
+
+// Status returns the HTTP Response Status Code for this authChallenge.
+func (ac *authChallenge) Status() int {
```
Comment (ncdc) [inline pkg/dockerregistry/auth/openshift.go]: I would remove this method and just return `http.StatusUnauthorized` directly in `ServeHTTP`

File: pkg/dockerregistry/auth/openshift.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+// challengeParams constructs the value to be used in
+// the WWW-Authenticate response challenge header.
+// See https://tools.ietf.org/html/rfc6750#section-3
+func (ac *authChallenge) challengeParams() string {
+	return fmt.Sprintf("Basic error=%s", ac.Error())
+}
+
+// SetHeader sets the WWW-Authenticate value for the given header.
+func (ac *authChallenge) SetHeader(header http.Header) {
```
Comment (ncdc) [inline pkg/dockerregistry/auth/openshift.go]: I would remove this method and set the header in `ServeHTTP`

File: pkg/dockerregistry/auth/openshift.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+// Status returns the HTTP Response Status Code for this authChallenge.
+func (ac *authChallenge) Status() int {
+	return http.StatusUnauthorized
+}
+
+// challengeParams constructs the value to be used in
+// the WWW-Authenticate response challenge header.
+// See https://tools.ietf.org/html/rfc6750#section-3
+func (ac *authChallenge) challengeParams() string {
+	return fmt.Sprintf("Basic error=%s", ac.Error())
```
Comment (ncdc) [inline pkg/dockerregistry/auth/openshift.go]: I think something like `Basic realm=openshift` would be more appropriate, although we may want the realm to be configurable.

File: pkg/dockerregistry/auth/openshift.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+	return &AccessController{
+		OpenShiftApiVersion: openshiftApiVersion.(string),
+		RegistryClient:      &rc,
+	}, nil
+}
+
+// Error returns the internal error string for this authChallenge.
+func (ac *authChallenge) Error() string {
+	return ac.err.Error()
```
Comment (ncdc) [inline pkg/dockerregistry/auth/openshift.go]: I don't think `ac.err` is ever set? If not, just remove this method.

Comment (ncdc) [inline pkg/dockerregistry/auth/openshift.go]: Nevermind, I see where it's set.

Comment (ncdc) [inline pkg/dockerregistry/auth/openshift.go]: However, I'm somewhat of the mindset to still say remove this, along with the various error definitions.

File: pkg/dockerregistry/auth/openshift.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+	authParts := strings.SplitN(req.Header.Get("Authorization"), " ", 2)
+	if len(authParts) != 2 || strings.ToLower(authParts[0]) != "basic" {
+		challenge.err = ErrTokenRequired
+		return nil, challenge
+	}
+	basicToken := authParts[1]
+
+	if len(accessRecords) == 0 {
+		return ctx, nil
```
Comment (ncdc) [inline pkg/dockerregistry/auth/openshift.go]: Why does this short circuit and return here?

File: pkg/dockerregistry/auth/openshift.go [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+	}
+
+	bearerToken := ""
+	for _, access := range accessRecords {
+		//fmt.Printf("%s:%s:%s", access.Resource.Type, access.Resource.Name, access.Action)
```
Comment (ncdc) [inline pkg/dockerregistry/auth/openshift.go]: Remove or convert to glog

File: pkg/dockerregistry/auth/openshift.go (hunk 4 above)
Comment (pravisankar) [inline pkg/dockerregistry/auth/openshift.go]: error definitions are used in the test cases

**Comment by pravisankar** [comment]
@ncdc tested the code (docker push/pull is working as expected)
I'll update the test cases, please take a look

**Comment by pravisankar** [comment]
@ncdc updated, please review

**Comment by ncdc** [comment]
Looks like you might need to run gofmt on test/integration/v2_docker_registry_test.go, according to travis.

File: pkg/dockerregistry/auth/openshift.go [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+
+var _ registryauth.AccessController = &AccessController{}
+
+type authChallenge struct {
+	err error
+}
+
+var _ registryauth.Challenge = &authChallenge{}
+
+type OpenShiftAccess struct {
```
Comment (ncdc) [inline pkg/dockerregistry/auth/openshift.go]: I'm not sure I like this being a struct. Could we remove it and modify `VerifyOpenShiftAccess` to take in the 4 fields as arguments instead?

File: pkg/dockerregistry/auth/openshift.go [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+// Error returns the internal error string for this authChallenge.
+func (ac *authChallenge) Error() string {
+	return ac.err.Error()
+}
+
+// challengeParams constructs the value to be used in
+// the WWW-Authenticate response challenge header.
+// See https://tools.ietf.org/html/rfc6750#section-3
+func (ac *authChallenge) challengeParams() string {
+	return fmt.Sprintf("Basic realm=openshift error=%s", ac.Error())
```
Comment (ncdc) [inline pkg/dockerregistry/auth/openshift.go]: The challenge response should not include "error=" in its text.

Comment (ncdc) [inline pkg/dockerregistry/auth/openshift.go]: I would remove this method and move the `Basic realm=...` down into ServeHTTP where you add the `WWW-Authenticate` header.

File: pkg/dockerregistry/auth/openshift.go [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+	return WithBearerToken(ctx, bearerToken), nil
+}
+
+func VerifyOpenShiftAccess(osAccess *OpenShiftAccess) error {
+	client, err := dockerregistry.NewUserOpenShiftClient(osAccess.BearerToken)
+	if err != nil {
+		return err
+	}
+	sar := authorizationapi.SubjectAccessReview{
+		TypeMeta: kapi.TypeMeta{
```
Comment (ncdc) [inline pkg/dockerregistry/auth/openshift.go]: You don't need TypeMeta

File: pkg/dockerregistry/auth/openshift.go [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+			APIVersion: "v1beta1",
+			Kind:       "SubjectAccessReview",
+		},
+		Verb:         osAccess.Verb,
+		Resource:     "imageRepositories",
```
Comment (ncdc) [inline pkg/dockerregistry/auth/openshift.go]: `imageRepositories` should be `imageStreams`

File: pkg/dockerregistry/auth/openshift_test.go [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+	"testing"
+
+	"github.com/docker/distribution/registry/auth"
+	"golang.org/x/net/context"
+)
+
+// TestVerifyOpenShiftAccess mocks openshift http request/response and
+// tests invalid/valid/scoped openshift tokens.
+func TestVerifyOpenShiftAccess(t *testing.T) {
+	// Test invalid openshift bearer token
```
Comment (ncdc) [inline pkg/dockerregistry/auth/openshift_test.go]: Could you please turn this into a table test?

File: pkg/dockerregistry/auth/openshift_test.go [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+	err = VerifyOpenShiftAccess(osAccess)
+	if err != nil {
+		t.Fatalf("expected nil for VerifyOpenShiftAccess but got %s", err)
+	}
+	server.Close()
+}
+
+// TestAccessController tests complete integration of the v2 registry auth package.
+func TestAccessController(t *testing.T) {
+	options := map[string]interface{}{
```
Comment (ncdc) [inline pkg/dockerregistry/auth/openshift_test.go]: Could you please turn this into a table test?

Comment (ncdc) [inline pkg/dockerregistry/auth/openshift_test.go]: Could you please add a test where the request is authorized?

File: pkg/dockerregistry/middleware/repository/openshift.go [hunk 13]
```diff
@@ -37,50 +38,20 @@ type repository struct {
...
-	}
-	registryClient, err := client.New(&registryClientConfig)
-	if err != nil {
-		return nil, fmt.Errorf("Error creating OpenShift client: %s", err)
-	}
-
 	nameParts := strings.SplitN(repo.Name(), "/", 2)
-
+	if len(nameParts) != 2 {
+		return nil, errors.New("Incorrect image repo name")
```
Comment (ncdc) [inline pkg/dockerregistry/middleware/repository/openshift.go]: s/repo/stream/

File: pkg/dockerregistry/middleware/repository/openshift.go [hunk 14]
```diff
@@ -239,32 +210,43 @@ func (r *repository) Put(manifest *manifest.SignedManifest) error {
...
+	if err != nil {
+		return nil, err
+	}
+	return client.ImageStreams(r.namespace).Get(r.name)
 }
 
 // getImage retrieves the Image with digest `dgst`. This uses the registry's
 // credentials and should ONLY
-func (r *repository) getImage(dgst digest.Digest) (*imageapi.Image, error) {
+func (r *repository) getImage(ctx context.Context, dgst digest.Digest) (*imageapi.Image, error) {
```
Comment (ncdc) [inline pkg/dockerregistry/middleware/repository/openshift.go]: `getImage` doesn't need `ctx` passed in

File: test/integration/v2_docker_registry_test.go [hunk 15]
```diff
@@ -191,9 +209,15 @@ middleware:
...
-func getTags(repoName string) ([]string, error) {
+func getTags(repoName, basicToken string) ([]string, error) {
 	url := fmt.Sprintf("http://127.0.0.1:5000/v2/%s/%s/tags/list", testutil.Namespace(), repoName)
-	resp, err := http.Get(url)
+	client := http.DefaultClient
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return []string{}, fmt.Errorf("error creating request: %v", err)
+	}
+	req.Header.Set("Authorization", "Basic "+basicToken)
```
Comment (ncdc) [inline test/integration/v2_docker_registry_test.go]: Please use http://golang.org/pkg/net/http/#Request.SetBasicAuth instead

File: test/integration/v2_docker_registry_test.go [hunk 16]
```diff
@@ -75,18 +77,26 @@ func signedManifest() ([]byte, digest.Digest, error) {
...
-		t.Errorf("unexpected error: %v", err)
+		t.Fatalf("error getting cluster admin client: %v", err)
 	}
-
 	clusterAdminClientConfig, err := testutil.GetClusterAdminClientConfig(clusterAdminKubeConfig)
 	if err != nil {
-		t.Errorf("unexpected error: %v", err)
+		t.Fatalf("error getting cluster admin client config: %v", err)
+	}
+	adminUser := "admin"
```
Comment (ncdc) [inline test/integration/v2_docker_registry_test.go]: Could you use a different user name besides admin here?

Comment (ncdc) [inline test/integration/v2_docker_registry_test.go]: Eh, nevermind

**Comment by pravisankar** [comment]
@ncdc updated as per your feedback, please review

File: pkg/dockerregistry/auth/openshift_test.go [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+				Action: "pull",
+			},
+			basicToken:          "b3BlbnNoaWZ0OmF3ZXNvbWU=",
+			openshiftStatusCode: 200,
+			openshiftResponse:   `{"namespace": "foo", "allowed": true, "reason": "authorized!", "kind": "SubjectAccessReviewResponse", "apiVersion": "v1beta1"}`,
+			expectedError:       nil,
+		},
+	}
+	for _, item := range table {
+		fmt.Println("Processing item: %s", item)
```
Comment (ncdc) [inline pkg/dockerregistry/auth/openshift_test.go]: Don't use println. t.Logf is ok

File: pkg/dockerregistry/auth/openshift_test.go [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+		req, err := http.NewRequest("GET", options["addr"].(string), nil)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if !item.skipBasicToken {
+			req.Header.Set("Authorization", fmt.Sprintf("Basic %s", item.basicToken))
```
Comment (ncdc) [inline pkg/dockerregistry/auth/openshift_test.go]: Please use http://golang.org/pkg/net/http/#Request.SetBasicAuth

File: pkg/dockerregistry/auth/openshift_test.go [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+	"testing"
+
+	"github.com/docker/distribution/registry/auth"
+	"golang.org/x/net/context"
+)
+
+// TestVerifyOpenShiftAccess mocks openshift http request/response and
+// tests invalid/valid/scoped openshift tokens.
+func TestVerifyOpenShiftAccess(t *testing.T) {
+	table := []struct {
```
Comment (ncdc) [inline pkg/dockerregistry/auth/openshift_test.go]: rename table to tests

File: pkg/dockerregistry/auth/openshift_test.go [hunk 20]
```diff
@@ -0,0 +1,189 @@
...
+			expectedError:       ErrOpenShiftAccessDenied,
+		},
+		{
+			// Test valid openshift bearer token and token scoped for create operation
+			openshiftStatusCode: 200,
+			openshiftResponse:   `{"namespace": "foo", "allowed": true, "reason": "authorized!", "kind": "SubjectAccessReviewResponse", "apiVersion": "v1beta1"}`,
+			expectedError:       nil,
+		},
+	}
+	for _, item := range table {
```
Comment (ncdc) [inline pkg/dockerregistry/auth/openshift_test.go]: rename item to test

File: pkg/dockerregistry/auth/openshift_test.go [hunk 21]
```diff
@@ -0,0 +1,189 @@
...
+	options := map[string]interface{}{
+		"addr":       "https://openshift-example.com/osapi",
+		"apiVersion": "v1beta1",
+	}
+	accessController, err := newAccessController(options)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	table := []struct {
```
Comment (ncdc) [inline pkg/dockerregistry/auth/openshift_test.go]: rename table to tests

File: pkg/dockerregistry/auth/openshift_test.go [hunk 22]
```diff
@@ -0,0 +1,189 @@
...
+				},
+				Action: "pull",
+			},
+			basicToken:          "b3BlbnNoaWZ0OmF3ZXNvbWU=",
+			openshiftStatusCode: 200,
+			openshiftResponse:   `{"namespace": "foo", "allowed": true, "reason": "authorized!", "kind": "SubjectAccessReviewResponse", "apiVersion": "v1beta1"}`,
+			expectedError:       nil,
+		},
+	}
+	for _, item := range table {
```
Comment (ncdc) [inline pkg/dockerregistry/auth/openshift_test.go]: rename item to test

File: pkg/dockerregistry/auth/openshift.go [hunk 23]
```diff
@@ -0,0 +1,159 @@
...
+func (ac *authChallenge) Error() string {
+	return ac.err.Error()
+}
+
+// ServeHttp handles writing the challenge response
+// by setting the challenge header and status code.
+func (ac *authChallenge) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// WWW-Authenticate response challenge header.
+	// See https://tools.ietf.org/html/rfc6750#section-3
+	w.Header().Add("WWW-Authenticate", fmt.Sprintf("Basic realm=openshift"))
```
Comment (ncdc) [inline pkg/dockerregistry/auth/openshift.go]: Sorry to make you change this again - I just read what Docker is doing. See e.g. https://github.com/docker/distribution/blob/master/registry/auth/token/accesscontroller.go#L106. Basically, always print `Basic realm=<realm>`, and optionally append `,error=<error string>` if necessary.

I also think the realm should be configurable, defaulting to `openshift` if nothing is specified.

File: pkg/dockerregistry/auth/openshift_test.go (hunk 18 above)
Comment (pravisankar) [inline pkg/dockerregistry/auth/openshift_test.go]: This was intentionally done to simulate invalid basic auth tokens

Comment (ncdc) [inline pkg/dockerregistry/auth/openshift_test.go]: Ok

**Comment by pravisankar** [comment]
@ncdc PTAL

**Comment by ncdc** [comment]
[test]

**Comment by ncdc** [comment]
[test]

**Comment by ncdc** [comment]
[test]

**Comment by ncdc** [comment]
LGTM [merge]

**Comment by ncdc** [comment]
exec failure, re [merge]

**Comment by ncdc** [comment]
exec flake, [merge]

**Comment by brenton** [comment]
Is there a link to a story or a short description somewhere of what this PR enables?  I'm guessing this will start enforcing who can push/pull images to the registry.

**Comment by brenton** [comment]
I see the policy is being updated to have a 'images' group.

**Comment by ncdc** [comment]
@brenton the card is here: https://trello.com/c/IdJ6y0Wt/295-5-registry-authentication-authorization-imageregistry-codefreeze

A user will be able to push to an image repository in the OpenShift registry if they have edit or admin rights to a project. For pull, view/edit/admin rights.

Once Docker 1.6 is out and we've switching to using the v2 registry, we'll be securing access to /images and /imageStreamMappings so that only the v2 registry may access those routes. Users will be allowed to access /imageStreams, /imageStreamTags, and /imageStreamImages. See #1652

**Comment by brenton** [comment]
Thanks Andy.  This is perfect.

**Pull request merged**